# Data Pipeline - PacCafe Exercise 1
**Author:** Naufal  
**Date:** 2026-04-28  
**Database:** PostgreSQL - paccafe  

## Overview
This notebook contains data profiling and quality checks 
for the PacCafe database.

In [124]:
import psycopg2
import pandas as pd 
from sqlalchemy import create_engine
import json
from sqlalchemy import text
from datetime import datetime

### Connecting data from database

In [125]:
engine = create_engine('postgresql+psycopg2://postgres:adminadmin@localhost:5435/paccafe')

## 1. Create a Function to List All Tables in the Database (5 Point)

* Objective: Create function for retrieving and displaying all the tables available in the connected database.

* Hint: You can user SQL query to access the list of the table

In [126]:
def list_all_tables(engine):
    with engine.connect() as conn:
        result = conn.execute(text("""
            SELECT table_name
            FROM information_schema.tables
            WHERE table_schema = 'public'
        """))
        tables = [row[0] for row in result]
    return tables

### Result
Successfully retrieved **6 tables** from the PacCafe database:
- products
- inventory_tracking
- orders
- order_details
- customers
- employees

## 2. Extract Data (5 Point)
- Objective: Create function to extract the data for tables in the database and return it in a structured format.
- Output: A dictionary where each key is a table name, and the corresponding value is a pandas DataFrame containing the data for that table. (Suggested )


coba casting untuk test hire_date jadi date_time

In [127]:
def extract_data(engine):
    tables = list_all_tables(engine)
    data = {}
    
    for table in tables:
        if table == 'employees':
            query = """
                SELECT 
                    employee_id,
                    first_name,
                    last_name,
                    CAST(hire_date AS TIMESTAMP) as hire_date,
                    role,
                    email,
                    created_at
                FROM employees
            """
        else:
            query = f"SELECT * FROM {table}"
        
        data[table] = pd.read_sql_query(query, engine)
    
    return data

# call it outside the function
data = extract_data(engine)
for table, df in data.items():
    print(f"'{table}': DataFrame")

'products': DataFrame
'inventory_tracking': DataFrame
'orders': DataFrame
'order_details': DataFrame
'customers': DataFrame
'employees': DataFrame


### Result
Successfully extracted all **6 tables** into DataFrames and stored 
in dictionary format:

| Table | Type |
|---|---|
| products | DataFrame |
| inventory_tracking | DataFrame |
| orders | DataFrame |
| order_details | DataFrame |
| customers | DataFrame |
| employees | DataFrame |

## 3. Create a Function to Get information for rows and cols (5 Point)
- objective: This function will retrieve the number of rows and columns for a specified table.
- Show the information number of rows and cols for all tables


In [128]:
def get_rows_cols(data, table_name=None):
    if table_name:
        # return for specific table
        df = data[table_name]
        return [df.shape[0], df.shape[1]]
    else:
        # original behavior - print all
        for table in data:
            df = data[table]
            rows = df.shape[0]
            cols = df.shape[1]
            print(f'{table} | Rows: {rows} | Cols: {cols}')

In [129]:
get_rows_cols(data)

products | Rows: 54 | Cols: 7
inventory_tracking | Rows: 162 | Cols: 6
orders | Rows: 1010 | Cols: 8
order_details | Rows: 3022 | Cols: 7
customers | Rows: 204 | Cols: 7
employees | Rows: 103 | Cols: 7


## 4. Create a Function to check the data types of all columns (5 Point)
- Objective: Create function that checks the data types of all columns in a specified table.
- Example Dictionary Format:


In [130]:
def get_cols_data_types(data, table_name=None):
    if table_name:
        # return for specific table
        df = data[table_name]
        return {col: str(df[col].dtype) for col in df.columns}
    else:
        # original behavior - all tables
        result = {}
        for table in data:
            df = data[table]
            result[table] = {col: str(df[col].dtype) for col in df.columns}
            print(f'{table} | Columns and Data Types:')
            print(df.dtypes)
            print('---')
        return result

In [131]:
# with specific table
print(get_cols_data_types(data, 'products'))

# without table name
get_cols_data_types(data)

{'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'unit_price': 'str', 'cost_price': 'str', 'in_stock': 'bool', 'created_at': 'datetime64[us]'}
products | Columns and Data Types:
product_id               int64
product_name               str
category                   str
unit_price                 str
cost_price                 str
in_stock                  bool
created_at      datetime64[us]
dtype: object
---
inventory_tracking | Columns and Data Types:
tracking_id                 int64
product_id                  int64
quantity_change             int64
change_date        datetime64[us]
reason                        str
created_at         datetime64[us]
dtype: object
---
orders | Columns and Data Types:
order_id                   int64
employee_id                int64
customer_id              float64
order_date        datetime64[us]
total_amount             float64
payment_method               str
order_status                 str
created_at        datetime64[us]
dtype:

{'products': {'product_id': 'int64',
  'product_name': 'str',
  'category': 'str',
  'unit_price': 'str',
  'cost_price': 'str',
  'in_stock': 'bool',
  'created_at': 'datetime64[us]'},
 'inventory_tracking': {'tracking_id': 'int64',
  'product_id': 'int64',
  'quantity_change': 'int64',
  'change_date': 'datetime64[us]',
  'reason': 'str',
  'created_at': 'datetime64[us]'},
 'orders': {'order_id': 'int64',
  'employee_id': 'int64',
  'customer_id': 'float64',
  'order_date': 'datetime64[us]',
  'total_amount': 'float64',
  'payment_method': 'str',
  'order_status': 'str',
  'created_at': 'datetime64[us]'},
 'order_details': {'order_detail_id': 'int64',
  'order_id': 'int64',
  'product_id': 'int64',
  'quantity': 'int64',
  'unit_price': 'float64',
  'subtotal': 'float64',
  'created_at': 'datetime64[us]'},
 'customers': {'customer_id': 'int64',
  'first_name': 'str',
  'last_name': 'str',
  'email': 'str',
  'phone': 'str',
  'loyalty_points': 'int64',
  'created_at': 'datetime64[us]

## 5. Create a Function to Check Unique Values in a Column (5 Point)
- Objective: Create function that checks the unique values present in a specific column of a specified table.
- Output: A list of unique values found in the column.
- Data Check:
    - Employees Table: Check unique values in the `role` column.  
    - Orders Table: Check unique values in the `payment_method` column.  
    - Products Table: Check unique values in the `category` column.  
    - Inventory Table: Check unique values in the `reason` column.

In [132]:
def check_unique_values(data, table_name, column_name=None):
    df = data[table_name]
    
    if column_name:
        # specific column - current behavior
        unique = df[column_name].unique().tolist()
        return {column_name: unique}
    else:
        # all columns - for report
        result = {}
        for col in df.columns:
            result[col] = df[col].unique().tolist()
        return result

In [133]:
print(check_unique_values(data, 'employees', 'role'))
print(check_unique_values(data, 'orders', 'payment_method'))
print(check_unique_values(data, 'products', 'category'))
print(check_unique_values(data, 'inventory_tracking', 'reason'))

{'role': ['Cashier', 'Waitress', 'Manager', 'Waiter', 'Barista', 'today', 'third', 'me']}
{'payment_method': ['Cash', 'Credit Card', 'Debit Card', 'ERROR']}
{'category': ['Pastry', 'Salad', 'Smoothie', 'Sandwich', 'Coffee', 'Tea']}
{'reason': ['Restock', 'Damaged', 'Expired', 'ERROR']}


In [134]:
# specific column - original behavior
print(check_unique_values(data, 'employees', 'role'))

# all columns - for report
print(check_unique_values(data, 'products'))

{'role': ['Cashier', 'Waitress', 'Manager', 'Waiter', 'Barista', 'today', 'third', 'me']}
{'product_id': [53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106], 'product_name': ['Pastry situation', 'Pastry sure', 'Pastry statement', 'Salad on', 'Salad really', 'Salad bag', 'Pastry security', 'Smoothie lead', 'Salad people', 'Sandwich role', 'Pastry plan', 'Sandwich international', 'Coffee commercial', 'Sandwich society', 'Smoothie TV', 'Sandwich room', 'Salad ahead', 'Salad east', 'Pastry raise', 'Pastry put', 'Smoothie around', 'Salad land', 'Coffee group', 'Salad truth', 'Sandwich although', 'Tea budget', 'Pastry subject', 'Sandwich sit', 'Pastry small', 'Pastry dream', 'Sandwich begin', 'Pastry how', 'Pastry glass', 'Coffee administration', 'Coffee ground', 'Salad American', 'Coffee citizen', 'Salad push', 'Sandwich 

### Result

| Table | Column | Unique Values | Issues Found |
|---|---|---|---|
| employees | role | Cashier, Waitress, Manager, Waiter, Barista, today, third, me | ⚠️ Invalid: 'today', 'third', 'me' |
| orders | payment_method | Cash, Credit Card, Debit Card, ERROR | ⚠️ Invalid: 'ERROR' |
| products | category | Pastry, Salad, Smoothie, Sandwich, Coffee, Tea | ✅ All valid |
| inventory_tracking | reason | Restock, Damaged, Expired, ERROR | ⚠️ Invalid: 'ERROR' |

### Finding
- 3 out of 4 checked columns contain **invalid values**
- Invalid values suggest **data entry errors** that need to be 
cleaned before analysis

## 6.Create Report (10 Point)
- Objective: Create a function to generate a data profiling report. in dictionary format and save it to a json file
- Output: json file

In [135]:
def generate_report(data, person_in_charge):
    report = {
        'person_in_charge': person_in_charge,
        'date_profiling': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
        'result': {}
    }
    
    for table in data:
        report['result'][table] = {
            'shape': get_rows_cols(data, table),
            'data_types': get_cols_data_types(data, table),
            'unique_values': check_unique_values(data, table)
    }
    
    # save to json file
    with open('profiling_report.json', 'w') as f:
        json.dump(report, f, indent=4, default=str)
    
    print('Report saved to profiling_report.json!')
    return report

### Checking Json File

In [136]:
report = generate_report(data, 'Naufal')

Report saved to profiling_report.json!


In [137]:
print(report['result'].keys())

dict_keys(['products', 'inventory_tracking', 'orders', 'order_details', 'customers', 'employees'])


report file stored in report.json file

## Data Quality
**Create functions to check data quality and build a report.**

## 1. Check Missing Value (10 Point)
- Objective: Count of missing values for each column.


In [138]:
def check_missing_values(data, table_name):
    df = data[table_name]
    result = {}
    for col in df.columns:
        n_data = len(df)
        n_missing = df[col].isnull().sum()
        percentage = n_missing * 100 / n_data
        percentage = round(percentage, 2)
        
        if n_missing > 0:
            result[col] = {
                'n_missing': str(n_missing),
                'percentage': float(percentage)
            }
    
    return result  

In [139]:
for table in data:
    print(f'\n{table}:')
    print(check_missing_values(data, table))


products:
{}

inventory_tracking:
{}

orders:
{'customer_id': {'n_missing': '250', 'percentage': 24.75}}

order_details:
{}

customers:
{'phone': {'n_missing': '4', 'percentage': 1.96}}

employees:
{}


Two columns with missing data:
Missing value : 
 1. Table orders, Column : customer_id (24.75%)
 2. Table customers, Column : phone = (1.96%)

### 2. Date Validation (10 Point)
- Obejective: Validate date columns for the format (e.g., yyyy-mm-dd, yyyy-mm-dd hh:mm:ss).

- Data Check
    - employees: hire_date (yyyy-mm-dd)
    - inventory_tracking: change_date (yyyy-mm-dd)
    - orders: order_date (yyyy-mm-dd hh:mm:ss)


data checking

In [140]:
print(data['employees']['hire_date'].head(5))
print(data['inventory_tracking']['change_date'].head(5))
print(data['orders']['order_date'].head(5))

0   2020-07-19
1   2023-05-15
2   2022-11-27
3   2021-07-06
4   2021-10-21
Name: hire_date, dtype: datetime64[us]
0   2024-11-01
1   2024-08-05
2   2023-11-12
3   2023-05-29
4   2023-09-22
Name: change_date, dtype: datetime64[us]
0   2023-12-04 15:41:20
1   2024-08-03 10:59:12
2   2023-12-13 12:11:13
3   2023-07-14 23:06:02
4   2024-12-21 10:33:32
Name: order_date, dtype: datetime64[us]


hire_date format still on string, next step we need to convert it to date_time in data cleaning proccess

In [141]:
def validate_dates(data, table_name, col_name, date_format):
    df = data[table_name]
    result = {
        'column': col_name,
        'is_datetime': False,
        'invalid_dates': []
    }
    
    if df[col_name].dtype == 'datetime64[us]':
        result['is_datetime'] = True
    else:
        result['is_datetime'] = False
        for value in df[col_name]:
            try:
                datetime.strptime(str(value), date_format)
            except:
                result['invalid_dates'].append(str(value))
    
    return result

In [142]:
print(validate_dates(data, 'employees', 'hire_date', '%Y-%m-%d'))
print(validate_dates(data, 'inventory_tracking', 'change_date', '%Y-%m-%d'))
print(validate_dates(data, 'orders', 'order_date', '%Y-%m-%d %H:%M:%S'))

{'column': 'hire_date', 'is_datetime': True, 'invalid_dates': []}
{'column': 'change_date', 'is_datetime': True, 'invalid_dates': []}
{'column': 'order_date', 'is_datetime': True, 'invalid_dates': []}


## Date Validation Summary

### Columns Checked
| Table | Column | Expected Format | Is Datetime | Invalid Values |
|---|---|---|---|---|
| employees | hire_date | yyyy-mm-dd | ✅ True | None |
| inventory_tracking | change_date | yyyy-mm-dd | ✅ True | None |
| orders | order_date | yyyy-mm-dd hh:mm:ss | ✅ True | None |


### Conclusion
- No invalid date formats found across all checked columns


## 3. Numeric Validation (10 Point)
- Objective: Check if columns contain valid numeric values.
- Data Check
    - products: unit_price, cost_price
    - orders: total_amount
    - order_details: unit_price, quantity, subtotal
    - inventory_tracking: quantity_change
    - customers: loyalty_points


### check numeric cols

In [143]:
numeric_cols = {
    'products': ['unit_price', 'cost_price'],
    'orders': ['total_amount'],
    'order_details': ['unit_price', 'quantity', 'subtotal'],
    'inventory_tracking': ['quantity_change'],
    'customers': ['loyalty_points']
}

for table, cols in numeric_cols.items():
    print(f'\n{table}:')
    for col in cols:
        print(f'  {col}: {data[table][col].dtype}')


products:
  unit_price: str
  cost_price: str

orders:
  total_amount: float64

order_details:
  unit_price: float64
  quantity: int64
  subtotal: float64

inventory_tracking:
  quantity_change: int64

customers:
  loyalty_points: int64


In [144]:
def validate_numeric(data, table_name, col_name):
    df = data[table_name]
    result = {
        'column': col_name,
        'is_numeric': False,
        'invalid_values': []
    }
    
    if pd.api.types.is_numeric_dtype(df[col_name]):
        result['is_numeric'] = True
    else:
        result['is_numeric'] = False
        # loop through values and find non-numeric ones
        for value in df[col_name]:
            try:
                float(value)  # try converting to number
            except:
                result['invalid_values'].append(str(value))
    
    return result

In [145]:
print(validate_numeric(data, 'products', 'unit_price'))
print(validate_numeric(data, 'products', 'cost_price'))
print(validate_numeric(data, 'orders', 'total_amount'))

{'column': 'unit_price', 'is_numeric': False, 'invalid_values': ['$47.00', '$47.00', '$17.00', '$21.00', '$43.00', '$9.00', '$2.00', '$48.00', '$10.00', '$13.00', '$10.00', '$37.00', '$18.00', '$47.00', '$37.00', '$33.00', '$41.00', '$23.00', '$29.00', '$11.00', '$23.00', '$37.00', '$39.00', '$42.00', '$32.00', '$19.00', '$32.00', '$15.00', '$35.00', '$12.00', '$23.00', '$21.00', '$34.00', '$44.00', '$7.00', '$21.00', '$6.00', '$20.00', '$10.00', '$33.00', '$48.00', '$23.00', '$13.00', '$7.00', '$34.00', '$46.00', '$20.00', '$35.00', '$4.00', '$38.00', '$19.00', '$37.00', '$-15.00', '$-11.00']}
{'column': 'cost_price', 'is_numeric': False, 'invalid_values': ['$43.00', '$43.00', '$14.00', '$20.00', '$40.00', '$1.00', '$-3.00', '$45.00', '$6.00', '$9.00', '$2.00', '$34.00', '$11.00', '$38.00', '$35.00', '$27.00', '$35.00', '$17.00', '$26.00', '$6.00', '$20.00', '$28.00', '$38.00', '$36.00', '$27.00', '$16.00', '$29.00', '$12.00', '$27.00', '$6.00', '$14.00', '$14.00', '$32.00', '$40.00',

In [146]:
print(validate_numeric(data, 'order_details', 'unit_price'))
print(validate_numeric(data, 'order_details', 'quantity'))
print(validate_numeric(data, 'order_details', 'subtotal'))
print(validate_numeric(data, 'inventory_tracking', 'quantity_change'))
print(validate_numeric(data, 'customers', 'loyalty_points'))

{'column': 'unit_price', 'is_numeric': True, 'invalid_values': []}
{'column': 'quantity', 'is_numeric': True, 'invalid_values': []}
{'column': 'subtotal', 'is_numeric': True, 'invalid_values': []}
{'column': 'quantity_change', 'is_numeric': True, 'invalid_values': []}
{'column': 'loyalty_points', 'is_numeric': True, 'invalid_values': []}


## Numeric Validation Summary

### Columns Checked
| Table | Column | Expected Type | Is Numeric | Invalid Values |
|---|---|---|---|---|
| products | unit_price | float64 | ❌ False | Contains `$` sign |
| products | cost_price | float64 | ❌ False | Contains `$` sign |
| orders | total_amount | float64 | ✅ True | None |
| order_details | unit_price | float64 | ✅ True | None |
| order_details | quantity | int64 | ✅ True | None |
| order_details | subtotal | float64 | ✅ True | None |
| inventory_tracking | quantity_change | int64 | ✅ True | None |
| customers | loyalty_points | int64 | ✅ True | None |

### Findings
- `unit_price` and `cost_price` in `products` table are stored as **string** instead of `float64`
  - Values contain `$` symbol (e.g. `$47.00`, `$-15.00`)
  - Need to remove `$` sign and convert to float during cleaning

### Conclusion
- 2 numeric type issues found in `products` table
- Both columns need cleaning: remove `$` and convert to `float64`
- Negative values found (`$-15.00`, `$-3.00`) — to be handled in next validation step

## 4. Negative Value Validation (10 Point)
- Objective: Identify any negative values in columns where they are not allowed.
- Data Check
    - products: unit_price, cost_price
    - orders: total_amount
    - order_details: unit_price, quantity, subtotal
    - inventory_tracking: quantity_change
    - customers: loyalty_points


In [147]:
print(data['inventory_tracking'].head())
print(data['inventory_tracking'].dtypes)

   tracking_id  product_id  quantity_change change_date   reason created_at
0            1          53                3  2024-11-01  Restock 2024-11-01
1            2          54               10  2024-08-05  Restock 2024-08-05
2            3          55                6  2023-11-12  Damaged 2023-11-12
3            4          56                8  2023-05-29  Damaged 2023-05-29
4            5          57                6  2023-09-22  Restock 2023-09-22
tracking_id                 int64
product_id                  int64
quantity_change             int64
change_date        datetime64[us]
reason                        str
created_at         datetime64[us]
dtype: object


In [148]:
def validate_negative(data, table_name, col_name):
    df = data[table_name]
    
    # handle non-numeric columns (like $ sign)
    if not pd.api.types.is_numeric_dtype(df[col_name]):
        values = df[col_name].str.replace('$', '').astype(float)
    else:
        values = df[col_name]
    
    negatives = values[values < 0]
    
    return {
        'column': col_name,
        'negative_count': len(negatives),
        'negative_values': negatives.tolist()
    }

In [149]:
print(validate_negative(data, 'products', 'unit_price'))
print(validate_negative(data, 'products', 'cost_price'))
print(validate_negative(data, 'orders', 'total_amount'))
print(validate_negative(data, 'order_details', 'unit_price'))
print(validate_negative(data, 'order_details', 'quantity'))
print(validate_negative(data, 'order_details', 'subtotal'))
print(validate_negative(data, 'customers', 'loyalty_points'))

{'column': 'unit_price', 'negative_count': 2, 'negative_values': [-15.0, -11.0]}
{'column': 'cost_price', 'negative_count': 3, 'negative_values': [-3.0, -3.0, -2.0]}
{'column': 'total_amount', 'negative_count': 0, 'negative_values': []}
{'column': 'unit_price', 'negative_count': 0, 'negative_values': []}
{'column': 'quantity', 'negative_count': 0, 'negative_values': []}
{'column': 'subtotal', 'negative_count': 0, 'negative_values': []}
{'column': 'loyalty_points', 'negative_count': 4, 'negative_values': [-710, -525, -304, -993]}


## Negative Value Validation Summary

### Columns Checked
| Table | Column | Can Be Negative | Negative Count | Values |
|---|---|---|---|---|
| products | unit_price | ❌ Never | 2 | -15.0, -11.0 |
| products | cost_price | ❌ Never | 3 | -3.0, -3.0, -2.0 |
| orders | total_amount | ❌ Never | 0 | None |
| order_details | unit_price | ❌ Never | 0 | None |
| order_details | quantity | ❌ Never | 0 | None |
| order_details | subtotal | ❌ Never | 0 | None |
| inventory_tracking | quantity_change | ✅ Allowed | - | Not checked |
| customers | loyalty_points | ❌ Never | 4 | -710, -525, -304, -993 |

### Business Rule: quantity_change Exception
`inventory_tracking.quantity_change` is intentionally excluded from negative validation because negative values are **valid by business logic**:
- `Restock` reason → positive values (items added)
- `Damaged` reason → negative values (items removed) ✅
- `Expired` reason → negative values (items removed) ✅

Negative values in this column represent inventory reduction which is a normal business operation.

### Findings
- `products.unit_price` → 2 negative prices found, likely data entry errors
- `products.cost_price` → 3 negative costs found, likely data entry errors
- `customers.loyalty_points` → 4 negative points found, customers cannot have negative loyalty points

### Conclusion
- 3 columns have invalid negative values that need to be fixed during cleaning
- Total invalid records: 9 rows across all tables

## 5. Build The Report (15 Point)
- Objective: Create a function to generate a data profiling report. in dictionary format and save it to a json file
- Output: json file

In [150]:
def generate_quality_report(data, person_in_charge):
    
    date_configs = {
        'employees': [('hire_date', '%Y-%m-%d')],
        'inventory_tracking': [('change_date', '%Y-%m-%d')],
        'orders': [('order_date', '%Y-%m-%d %H:%M:%S')]
    }
    
    numeric_configs = {
        'products': ['unit_price', 'cost_price'],
        'orders': ['total_amount'],
        'order_details': ['unit_price', 'quantity', 'subtotal'],
        'inventory_tracking': ['quantity_change'],
        'customers': ['loyalty_points']
    }
    
    negative_configs = {
        'products': ['unit_price', 'cost_price'],
        'orders': ['total_amount'],
        'order_details': ['unit_price', 'quantity', 'subtotal'],
        'customers': ['loyalty_points']
    }
    
    report = {
        'person_in_charge': person_in_charge,
        'date_quality_check': datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
        'result': {}
    }
    
    for table in data:
        report['result'][table] = {
            'missing_values': check_missing_values(data, table),
            'date_validity': {},
            'numeric_validity': {},
            'negative_validity': {}
        }
        
        if table in date_configs:
            for col, date_format in date_configs[table]:
                report['result'][table]['date_validity'][col] = validate_dates(data, table, col, date_format)
        
        if table in numeric_configs:
            for col in numeric_configs[table]:
                report['result'][table]['numeric_validity'][col] = validate_numeric(data, table, col)
        
        if table in negative_configs:
            for col in negative_configs[table]:
                report['result'][table]['negative_validity'][col] = validate_negative(data, table, col)
    
    # save to json
    with open('quality_report.json', 'w') as f:
        json.dump(report, f, indent=4, default=str)
    
    print('Quality report saved to quality_report.json!')
    return report

quality_report = generate_quality_report(data, 'Naufal')

Quality report saved to quality_report.json!


## Recommendations

Based on data profiling and quality checks, several cleaning processes 
are needed to make the dataset consumable by analysts.

### 1. Fix Numeric Columns with $ Sign (products)
**Affected columns:** `unit_price`, `cost_price`  
**Issue:** Both columns are stored as string due to `$` symbol, contains negative values  
**Why:** Numeric columns stored as strings cannot be used in calculations, 
aggregations, or financial analysis. Negative prices are logically impossible 
for a cafe menu and indicate data entry errors.  
**Recommendation:**
- Remove `$` symbol from all values and convert to `float64`
- Flag negative values for manual review — price data is sensitive 
  and should not be auto-fixed without human approval
- Add constraint: `price >= 1` to prevent invalid prices in future

### 2. Flag Invalid Role Values (employees)
**Affected column:** `role`  
**Issue:** Invalid values found: `'today'`, `'third'`, `'me'`  
**Why:** Invalid roles can cause incorrect access control, reporting errors, 
and HR data inconsistencies. Only predefined roles should exist in the system.  
**Recommendation:**
- Flag records with invalid roles for manual review
- Cannot assume correct role without business confirmation
- Add constraint: `role IN ('Cashier', 'Waitress', 'Manager', 'Waiter', 'Barista')`
  to prevent invalid roles from being inserted in future

### 3. Handle Missing Customer ID (orders)
**Affected column:** `customer_id`  
**Issue:** 250 records (24.75%) have no customer linked  
**Why:** Missing customer IDs break customer purchase analysis, loyalty tracking, 
and revenue attribution. Customer ID is a mandatory field for order records.  
**Recommendation:**
- Assign `'UNKNOWN'` for existing missing values — likely walk-in customers
- Add `NOT NULL` constraint to prevent future orders without customer reference

### 4. Handle ERROR Values (orders, inventory_tracking)
**Affected columns:** `payment_method`, `reason`  
**Issue:** Invalid value `'ERROR'` found in both columns  
**Why:** ERROR values pollute analysis and reporting. Payment method affects 
financial reconciliation while inventory reason affects stock management accuracy.  
**Recommendation:**
- Replace `'ERROR'` with `null` to preserve row data
- Add constraints to allow only valid values:
  - `payment_method IN ('Cash', 'Credit Card', 'Debit Card')`
  - `reason IN ('Restock', 'Damaged', 'Expired')`

### 5. Handle Missing Phone Numbers (customers)
**Affected column:** `phone`  
**Issue:** 4 records (1.96%) missing phone number  
**Why:** Phone number is an optional field — customers may choose not to provide it. 
Low percentage makes it non-critical.  
**Recommendation:**
- Leave as `null` — phone is an optional field
- No constraint needed as null should be allowed

### 6. Fix Negative Loyalty Points (customers)
**Affected column:** `loyalty_points`  
**Issue:** 4 records have negative points (-710, -525, -304, -993)  
**Why:** Negative loyalty points are logically impossible — customers earn 
points through purchases and cannot go below zero. These indicate data entry errors.  
**Recommendation:**
- Flag for manual review before applying `abs()` to convert to positive
- Add constraint: `loyalty_points >= 0` to prevent negative points in future

---

## Constraint Implementation (SQL Reference)

> **NOTES:** These SQL statements should only be executed AFTER 
> existing data violations are cleaned. Running constraints on dirty data 
> will result in PostgreSQL rejecting the constraint.

### Products - Price Constraint
```sql
ALTER TABLE products
ADD CONSTRAINT check_unit_price CHECK (unit_price >= 1);

ALTER TABLE products
ADD CONSTRAINT check_cost_price CHECK (cost_price >= 1);
```

### Employees - Role Constraint
```sql
ALTER TABLE employees
ADD CONSTRAINT check_role 
CHECK (role IN ('Cashier', 'Waitress', 'Manager', 'Waiter', 'Barista'));
```

### Orders - Customer ID NOT NULL
```sql
ALTER TABLE orders
ALTER COLUMN customer_id SET NOT NULL;
```

### Orders - Payment Method Constraint
```sql
ALTER TABLE orders
ADD CONSTRAINT check_payment_method 
CHECK (payment_method IN ('Cash', 'Credit Card', 'Debit Card'));
```

### Inventory Tracking - Reason Constraint
```sql
ALTER TABLE inventory_tracking
ADD CONSTRAINT check_reason 
CHECK (reason IN ('Restock', 'Damaged', 'Expired'));
```